subprojects/internal-performance-testing/src/main/groovy/org/gradle/performance/results/IndexPageGenerator.java

```diff
@@ -16,96 +16,238 @@
 
 package org.gradle.performance.results;
 
-import org.gradle.api.Transformer;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
 import org.gradle.performance.measure.DataSeries;
 import org.gradle.performance.measure.Duration;
+import org.gradle.performance.util.Git;
 
+import java.io.File;
 import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.io.Writer;
-import java.util.Date;
-import java.util.LinkedHashMap;
+import java.util.Comparator;
 import java.util.List;
-import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static java.util.Comparator.comparing;
+import static java.util.Comparator.comparingDouble;
 
 public class IndexPageGenerator extends HtmlPageGenerator<ResultsStore> {
-    private final List<NavigationItem> navigationItems;
+    public static final int DANGEROUS_REGRESSION_CONFIDENCE_THRESHOLD = 90;
+    private final Set<ScenarioBuildResultData> scenarios;
+    private final ResultsStore resultsStore;
+    private final String commitId = Git.current().getCommitId();
+
+    public IndexPageGenerator(ResultsStore resultsStore, File resultJson) {
+        this.resultsStore = resultsStore;
+        this.scenarios = readBuildResultData(resultJson);
+    }
+
+    private Set<ScenarioBuildResultData> readBuildResultData(File resultJson) {
+        try {
+            List<ScenarioBuildResultData> list = new ObjectMapper().readValue(resultJson, new TypeReference<List<ScenarioBuildResultData>>() {
+            });
+            return sortBuildResultData(list.stream().map(this::queryExecutionData));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    @VisibleForTesting
+    Set<ScenarioBuildResultData> sortBuildResultData(Stream<ScenarioBuildResultData> data) {
+        Comparator<ScenarioBuildResultData> comparator = comparing(ScenarioBuildResultData::isBuildFailed).reversed()
+            .thenComparing(ScenarioBuildResultData::isSuccessful)
+            .thenComparing(comparing(ScenarioBuildResultData::isAboutToRegress).reversed())
+            .thenComparing(comparingDouble(ScenarioBuildResultData::getConfidencePercentage).reversed())
+            .thenComparing(comparingDouble(ScenarioBuildResultData::getRegressionPercentage).reversed())
+            .thenComparing(ScenarioBuildResultData::getScenarioName);
+        return data.collect(() -> new TreeSet<>(comparator), TreeSet::add, TreeSet::addAll);
+    }
+
+    private ScenarioBuildResultData queryExecutionData(ScenarioBuildResultData scenario) {
+        PerformanceTestHistory history = resultsStore.getTestResults(scenario.getScenarioName(), 3, 2, ResultsStoreHelper.determineChannel());
```

Comment:
We use `3` here, since this is the number of retries? If that is the case, let's extract a constant with that name.

Where does the restriction on the last `2` days come from?

Reply:
Yes, 3 is the retry count. 2 means only fetch data from last two days. I was not sure what's the proper number here, but since we just care the execution of "current commit", I thought 2 days is enough to fetch "current commit" data. WDYT? If 2 days is not good, then what this number should be?

Reply:
2 days is probably good enough. It should still be a constant with a speaking name I think.

Reply:
Extracted to a constant.